### PR TITLE
Some performance optimizations

### PR DESCRIPTION
**Pull request by ovk**
While playing  with the library I noticed that performance quickly degrades to abysmal levels with higher resolution: even on Full HD it can't hold 60 FPS on my hardware. Looking at profiling and tracing data revealed two main issues, which I addressed in this PR.

One issue was with floor level image being allocated on each frame, putting huge pressure on GC. It was addressed by using same image buffer for all frames, and writing it to screen with `ReplacePixels`, which seems to be working fine.

The other issue was due to number of goroutines growing out of control with increased resolution. This one I solved by simply using fixed number of goroutines for walls and sprites (100 and 32 respectively, as these seems to provide the best performance in my setup). I also removed some unnecessary nesting of goroutines and semaphore channel.

After these changes, FPS in my game increased by about 55% (although, this depends on particular scene being used for test).

Feel free to rewrite the code to your liking, if you take this PR. In either case, thank you for this fantastic library which is easy and fun to use.

### CODE REVIEW

**Comment by harbdog** [comment]
Thanks for the PR, I'll look over and play with it tomorrow!

File: render.go [hunk 1]
```diff
@@ -12,6 +11,9 @@ import (
 func (c *Camera) Draw(screen *ebiten.Image) {
 	screen.Clear()
 
+	// draw textured floor
+	screen.ReplacePixels(c.floorLvl.horBuffer.Pix)
```
Comment (harbdog) [inline render.go:15]: I can take care of this:

`// Deprecated: as of v2.4. Use WritePixels instead.`
https://github.com/hajimehoshi/ebiten/blob/d31b0189a20dbc3db28ba814774533a22fef1c17/image.go#L977

**Comment by harbdog** [comment]
@ovk it appears that this change has disabled floor texture rendering, which could be one reason why it might be so much faster. I'll try to poke around the changes to see what it takes to get it back.

from the [raycaster-go-demo](https://github.com/harbdog/raycaster-go-demo) project:

`go get github.com/harbdog/raycaster-go@ovk_performance`

image

**Comment by ovk** [comment]
Very strange. This [game](https://ovk.itch.io/siegfrost) uses raycaster-go from my branch and as you can see floor texturing works just fine, unless I'm misunderstanding something.

**Comment by harbdog** [comment]
Funny game! Well done getting it to run in browser too!

@ovk I figured it out, your game is likely not setting the optional simple floor texture like the demo project does. The render order still has it drawn after the raycast textured floor render so its getting overridden. I should be able to sort this out, thanks again!

```golang
        screen.Clear()

	// draw textured floor
	screen.WritePixels(c.floorLvl.horBuffer.Pix)

	//--draw basic sky and floor--//
	texRect := image.Rect(0, 0, c.texSize, c.texSize)
	lightingRGBA := &color.RGBA{R: c.maxLightRGB.R, G: c.maxLightRGB.G, B: c.maxLightRGB.B, A: 255}

	floorRect := image.Rect(0, int(float64(c.h)*0.5)+c.pitch,
		c.w, c.h)
	drawTexture(screen, c.floor, &floorRect, &texRect, lightingRGBA)
```

File: camera.go [hunk 2]
```diff
@@ -265,38 +255,66 @@ func (c *Camera) raycast() {
...
+	stride := numSprites / 32
+	if stride < 1 {
+		stride = 1
+	}
 
-	c.semaphore <- struct{}{} // Lock
-	defer func() {
-		<-c.semaphore // Unlock
-	}()
+	for i := 0; i < numSprites; i++ {
```
Comment (harbdog) [inline camera.go:297]: should be `for i := 0; i < numSprites; i += stride {`

I'll go ahead and fix this in the ovk_performance branch prior to getting it ready to merge to main.

Comment (ovk) [reply camera.go:297]: Yes, it should be, good catch.

**Comment by ovk** [comment]
Thanks @harbdog . 
You are right, I didn't use simple floor texture, so I guess that's why I didn't see the issue. Hopefully, changing the draw order should fix it.

**Comment by harbdog** [comment]
In another game I'm making I actually need the simple floor texture to render under the raycasted floor at a certain distance away, and so I'll be making another adjustment to make it so the simple texture only draws over areas where there's nothing being rendered from the raycasted floor. Also will have to clear the raycasted floor image each update to prevent ghost images from previous frames (because I have large open areas without walls).

**Comment by harbdog** [comment]
@ovk out of curiosity what are the specs you are running to see the 55% gains in performance? What OS, CPU, RAM and GPU?

**Comment by ovk** [comment]
My setup is i7 4700MQ with integrated GPU, 16Gb RAM, Arch with Sway. At 1280x720 resolution FPS was just a bit over 60, and after the changes it went up to ~95. I also asked a friend to test and he had ~120 -> ~170 FPS improvement on Ryzen 3200G, Manjaro with X11 session. The experience also became much smoother in general, but this is hard to quantify.

The most dramatic change was in WASM (Web) version, and this is why I started looking at pprof data in the first place. The original version was practically unplayable for me, with under 15FPS at 800x600 resolution and constant freezes. Updated versions runs close to 30FPS at 960x540 and is much smoother. Mu guess is that jerkiness and freezes were the result of GC having to run very frequently. I ran it with `GODEBUG=gctrace=1` before and after, and frequency of GC went down by about an order of magnitude.

**Comment by harbdog** [comment]
Sorry now back to looking at this, I've got a much newer CPU/GPU (i7 10700k, RTX 3800) which isn't seeing any FPS improvements, and in some cases a bit less FPS for higher resolutions (at least in the game I'm making which has huge maps without as many walls). I might need to tune it a bit, perhaps that "100" value for the striping needs to be more dynamic so that newer CPUs with more cores/threads can take advantage.

**Comment by ovk** [comment]
Makes sense. Perhaps, these values should be derived from `GOMAXPROCS` value.
Also, the main performance improvement was from removing those extra allocations each frame.

**Comment by harbdog** [comment]
I played around with the "100" value some, lower and higher, but none really made it noticeably faster. So instead I pulled out my old laptop which is somewhat comparable, with an i3 M330 and Rv710 running Ubuntu. Using the raycaster-go-demo I noticed no difference in framerate (20-25 fps at 1024x768) from the existing code. Can you try it out too? Also worth noting I'm using Golang version 1.17.x

```bash
git clone https://github.com/harbdog/raycaster-go-demo.git
```

first from the main branch:

```bash
go run main.go
```

then from the `ovk_performance` branch:

```bash
git checkout ovk_performance
go run main.go
```

**Comment by ovk** [comment]
I can see noticeable improvement. In your scene, I noticed that frame rate is very dependent on the part of scene camera looks at, so I added a printf to dump framerate into a file, as I walked around the scene, and I then used LibreOffice Calc to plot this simple graph comparing framerates of main and ovk_performance branches (VScync disabled in both): 

20221227_21h28m25s_grim

**Comment by harbdog** [comment]
Here's where I'm seeing quite the opposite, to reduce differences of DirectX vs. OpenGL I tried this time on my newer laptop running Ubuntu (i7-1065G7 with Iris G7):

main branch scene (~77-80 fps):
main_scene

ovk_performance scene (~52-55 fps):
ovk_performance_scene

**Comment by harbdog** [comment]
In case it matters, what version of golang are you using?

**Comment by ovk** [comment]
It's Go 1.19.4, and I also compiled the code (with `go build`) before doing these measurements.

I took a brief look at the code, and it looks like you put the frame-sized buffer allocation back (which happens each frame in `camera.Update()`: `c.floorLvl.initialize(c.w, c.h)`). As I said before, this leads to quite significant performance degradation and puts additional load on GC (see top of allocations from pprof below, all samples are pretty much just this one allocation).
With this allocation back, the only other difference is the number of goroutines, which I guess can lead to small performance boost or degradation, depending on the CPU. On my hardware, the sweet spot was ~100, and increasing/decreasing them led to FPS drop by about 5-8%. Seems like on your hardware the optimal value is different.

20221228_08h58m03s_grim

**Comment by harbdog** [comment]
I see now, so I might have to make a flag for that feature (unless I can figure a better way to clear those pixels), because if I don't reinitialize the buffer when there's no walls within the render distance it causes the floor texture to leave around previously drawn junk, which looks like this:

image


When the buffer is reinitialized each update it correctly draws the basic floor texture beneath without the junk textures:
image

**Comment by ovk** [comment]
Maybe keep the object, but clear image contents itself (e.g. by writing transparent or sky-colored pixels there). This way there won't be an allocation/disposal of the whole buffer each frame.

**Comment by harbdog** [comment]
Clearing the individual pixels was slower than just re-initializing it, so I've setup an optional flag to have it cleared (aka re-initialized) during updates. By default it will be false to not clear it during updates.

https://github.com/harbdog/raycaster-go/commit/3fcca82debebe756752bd02510e7c5f8bfee2084

**Comment by harbdog** [comment]
@ovk FYI I've started enhancement issue #28 to continue any further discussions on getting this merged into the main branch

**Pull request merged**